src/deserialize/parse.rs

```diff
@@ -0,0 +1,361 @@
+use crate::edn::{Edn, Error, List, Map, Set, Vector};
+
+pub(crate) fn tokenize(edn: &str) -> std::str::Chars {
+    edn.chars()
+}
+
+pub(crate) fn parse(c: Option<char>, chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    Ok(match c {
+        Some('[') => read_vec(chars)?,
+        Some('(') => read_list(chars)?,
+        Some('@') => read_set(chars)?,
+        Some('{') => read_map(chars)?,
+        edn => parse_edn(edn, chars)?,
+    })
+}
+
+pub(crate) fn parse_edn(c: Option<char>, chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    match c {
+        Some('\"') => Ok(read_str(chars)),
+        Some(':') => Ok(read_key(chars)),
+        Some(n) if n.is_numeric() => Ok(read_number(n, chars)?),
+        Some('-') => Ok(read_number('-', chars)?),
+        Some('\\') => Ok(read_char(chars)?),
+        Some(b) if b == 't' || b == 'f' || b == 'n' => Ok(read_bool_or_nil(b, chars)?),
+        a => Err(Error::ParseEdn(format!(
+            "{} could not be parsed",
+            a.unwrap().to_string()
```

Comment:
Is this sure to be an `Ok` value?

Maybe it is more verbose, but we can divide this match arm into two, one where `a` is `Ok(...)` and other where it is `Err(...)`.

Then the `unwrap` is unnecessary and we can use the error that came from `Err`

Reply:
abd451e